### DISCUSSION TITLE
Pass data to signIn callback

### DISCUSSION THREAD

**Discussion by callumok2004** [nextauthjs/next-auth#2703] [Help]
How can I pass data to the sign in function response?
This would mean I can attempt to authorize, and if there is an incorrect username/password or some other issue I can get that from the response.

**Comment by balazsorban44** [comment]
what do you mean by sign in function response exactly?

Do you have a code example/better description of what you are trying to achieve?

**Reply by callumok2004** [reply]
I mean this function below
```tsx
import { signIn } from 'next-auth/client';

// ...

await signIn("credentials", {
  redirect: false,
  email: $("input[type=email]").val(),
  password: $("input[type=password]").val(),
});
```

I'd like to get a better response from the promise, which shows if the sign in attempt failed or not, and if it did fail, it would be nice if I could pass an error message to it, or if there is a better way of doing what I am trying to achieve completely?

**Reply by balazsorban44** [reply]
Have you seen the docs? https://next-auth.js.org/getting-started/client#using-the-redirect-false-option

The error message will be available as `const {error} = await signIn(...)`

**Reply by callumok2004** [reply]
I cannot customise the error message generated by that response, which is what I want to do.

**Reply by balazsorban44** [reply]
I would guess you wanted to provide info like if their user or password was incorrect, instead of only getting `CredentialsSignin` error.

Ref: https://next-auth.js.org/configuration/pages#sign-in-page

You generally don't want to expose the EXACT reason why the login failed, because that information can be used by malicious attackers. For example, if you say the e-mail was incorrect, they could actually get a list of all your user's email addresses just by brute-forcing.

Make the user double-check all their credentials info instead.

**Comment by neilkinnish** [comment]
@balazsorban44 is it possible to get hold of the "callbackUrl" within the signIn callback function?

We are doing a check to see if the user has been invited in the function and if not disallow them to continue. If they are already signed up or have an invite they can continue... but we would like to know the URL they were intending to go to (not to use as we know we can use the callback function for that, but just to cross reference with the invite data).

**Reply by balazsorban44** [reply]
I believe it's available as `req.body.callbackUrl`, if you pack it differently:

```js
export default async function auth(req, res) {
  return await NextAuth(req, res, {
    callbacks: {
      signIn() {
        const { callbackUrl } = req.body
      }
    }
  })
}
```

Don't redirect to that in the `signIn` function, as it hasn't been verified by `next-auth` at that point.
